File: cms/src/api/shows.ts
/**
 * Shows API surface (Phase 7B).
 *
 * Mirrors `backend/app/api/shows.py`. Every call goes through the
 * centralised `apiFetch` so token + error handling stay consistent.
 */
import { apiFetch } from './client';
import type {
  Paginated,
  Show,
  ShowCreateInput,
  ShowSection,
  ShowStatus,
  ShowUpdateInput,
} from '../types/api';

export interface ListShowsParams {
  search?: string;
  /** Filter by show status enum. */
  status?: ShowStatus;
  /** Filter by platform section. */
  section?: ShowSection;
  page?: number;
  page_size?: number;
}

/** GET /api/shows — paginated list of shows. */
export function listShows(
  params: ListShowsParams = {},
): Promise<Paginated<Show>> {
  return apiFetch<Paginated<Show>>('/api/shows', {
    query: {
      search: params.search,
      status: params.status,
      section: params.section,
      page: params.page,
      page_size: params.page_size,
    },
  });
}

/** GET /api/shows/{id} — fetch a single show. */
export function getShow(showId: number): Promise<Show> {
  return apiFetch<Show>(`/api/shows/${showId}`);
}

/** POST /api/shows — create a new show. */
export function createShow(payload: ShowCreateInput): Promise<Show> {
  return apiFetch<Show>('/api/shows', {
    method: 'POST',
    body: payload,
  });
}

/** PUT /api/shows/{id} — partial update. */
export function updateShow(
  showId: number,
  payload: ShowUpdateInput,
): Promise<Show> {
  return apiFetch<Show>(`/api/shows/${showId}`, {
    method: 'PUT',
    body: payload,
  });
}

/** DELETE /api/shows/{id} — delete a show. */
export function deleteShow(showId: number): Promise<void> {
  return apiFetch<void>(`/api/shows/${showId}`, {
    method: 'DELETE',
  });
}

File: backend/app/api/shows.py
"""
Show CRUD endpoints.

    GET    /api/shows              - paginated list
    POST   /api/shows              - create
    GET    /api/shows/{show_id}    - retrieve
    PUT    /api/shows/{show_id}    - partial update
    DELETE /api/shows/{show_id}    - delete

Query parameters for list:
    search        : case-insensitive substring on title
    status        : filter by show status
    section       : filter by platform section
    page          : 1-based page number (default 1)
    page_size     : items per page (default 20, max 100)

All endpoints require an authenticated editor or admin (Phase 5).
"""

from typing import Annotated

from fastapi import APIRouter, Query, Response, status

from app.api.deps import DBSession, EditorDep
from app.core.enums import SECTIONS, is_valid_section
from app.core.pagination import clamp_page, clamp_page_size
from app.models.show import ShowStatus
from app.schemas.show import ShowCreate, ShowRead, ShowUpdate
from app.services import shows as show_service
from app.services.errors import ValidationFailure


router = APIRouter(prefix="/api/shows", tags=["shows"])


@router.get("")
def list_shows(
    db: DBSession,
    _user: EditorDep,
    search: Annotated[str | None, Query(max_length=255)] = None,
    status: Annotated[ShowStatus | None, Query()] = None,
    section: Annotated[str | None, Query(max_length=32)] = None,
    page: Annotated[int, Query(ge=1)] = 1,
    page_size: Annotated[int, Query(ge=1, le=100)] = 20,
):
    """Paginated list of shows with optional filters."""
    page = clamp_page(page)
    page_size = clamp_page_size(page_size)

    if section is not None and not is_valid_section(section):
        raise ValidationFailure(
            f"section must be one of {SECTIONS}",
            code="invalid_section",
            field="section",
        )

    items, total = show_service.list_shows(
        db,
        page=page,
        page_size=page_size,
        search=search,
        status=status,
        section=section,
    )
    return {
        "items": [ShowRead.model_validate(s) for s in items],
        "page": page,
        "page_size": page_size,
        "total": total,
    }


@router.post("", response_model=ShowRead, status_code=status.HTTP_201_CREATED)
def create_show(payload: ShowCreate, db: DBSession, _user: EditorDep):
    """Create a new show. Returns 409 on slug conflict."""
    return show_service.create_show(db, payload)


@router.get("/{show_id}", response_model=ShowRead)
def get_show(show_id: int, db: DBSession, _user: EditorDep):
    """Retrieve a show by id. Returns 404 if absent."""
    return show_service.get_show(db, show_id)


@router.put("/{show_id}", response_model=ShowRead)
def update_show(show_id: int, payload: ShowUpdate, db: DBSession, _user: EditorDep):
    """Partial update. Returns 404 / 409 on errors."""
    return show_service.update_show(db, show_id, payload)


@router.delete("/{show_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_show(show_id: int, db: DBSession, _user: EditorDep):
    """Delete a show. Returns 204 on success, 404 if absent."""
    show_service.delete_show(db, show_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


__all__ = ["router"]